src/module_2/eda1.py

```diff
@@ -0,0 +1,206 @@
+
+
+def descriptive_analysis(df: pd.DataFrame, name: str) -> None:
+    logger.info(f"\nDescriptive Analysis for {name}")
+    try:
+        logger.info(df.describe(include="all"))
+    except Exception as e:
+        logger.error(f"Error performing descriptive analysis on {name}: {e}")
+
+
+for df, name in zip(
+    [orders, regulars, abandoned_cart, inventory, users],
+    ["Orders", "Regulars", "Abandoned Cart", "Inventory", "Users"],
+):
+    descriptive_analysis(df, name)
+
+sns.set(style="whitegrid")
+
+
+def plot_item_distribution(df: pd.DataFrame, name: str) -> None:
+    if "item_ids" in df.columns:
+        try:
+            plt.figure(figsize=(10, 6))
+            sns.histplot(df["item_ids"].apply(len), bins=30, kde=True)
+            plt.title(f"Distribution of Number of Items in {name}")
+            plt.xlabel("Number of Items")
+            plt.ylabel("Frequency")
+            plt.show()
+        except Exception as e:
+            logger.error(f"Error plotting item distribution for {name}: {e}")
+    else:
+        logger.warning(f"The column 'item_ids' is not found in {name}.")
+
+
+plot_item_distribution(orders, "Orders")
+plot_item_distribution(abandoned_cart, "Abandoned Carts")
+
+try:
+    regular_users = set(regulars["user_id"])
+    abandoned_cart["regular_user"] = abandoned_cart["user_id"].isin(regular_users)
+
+    abandonment_rates = abandoned_cart.groupby("regular_user").size() / users.shape[0]
+    logger.info("Abandonment Rate of Carts by User Type")
+    logger.info(abandonment_rates)
+
+    abandonment_rates = abandonment_rates.reset_index(name="abandonment_rate")
+    abandonment_rates["regular_user"] = abandonment_rates["regular_user"].replace(
+        {True: "Regular", False: "Non-Regular"}
+    )
+
+    plt.figure(figsize=(8, 5))
+    sns.barplot(x="regular_user", y="abandonment_rate", data=abandonment_rates)
+    plt.title("Abandonment Rate of Carts by User Type")
+    plt.xlabel("User Type")
+    plt.ylabel("Abandonment Rate")
+    plt.ylim(0, 1)
+    plt.show()
+except Exception as e:
+    logger.error(f"Error calculating abandonment rates: {e}")
+
+try:
+    if "total_amount" in orders.columns:
+        avg_total_amount = orders["total_amount"].mean()
+        logger.info(f"The average total amount of orders is: {avg_total_amount}")
+
+        abandoned_cart["created_at"] = pd.to_datetime(abandoned_cart["created_at"])
+        abandoned_cart["abandon_date"] = abandoned_cart["created_at"].dt.date
+
+        abandonment_by_date = abandoned_cart.groupby("abandon_date").size()
+
+        plt.figure(figsize=(12, 6))
+        sns.lineplot(x=abandonment_by_date.index, y=abandonment_by_date.values)
+        plt.title("Number of Abandoned Carts by Date")
+        plt.xlabel("Date")
+        plt.ylabel("Number of Abandoned Carts")
+        plt.xticks(rotation=45)
+        plt.show()
+    else:
+        logger.warning("Cannot perform alternative analysis: missing required columns.")
```

Comment:
What you have done here is good, but I'm missing more depth in the analysis. How's top customers and top items interact? Are cohorts of customers that are more prone to some products? Do top customers interact uniformly the same over time, or do they have peaks of engagement? What about products, do they get trendy over time or are they constant? What are the most common items in abandonned carts that are not in regular carts? Etc.